### PR TITLE
refactor: remove getter on date fields in mongoose models

### LINKED ISSUE #207: Remove getter on Date fields in mongoose models
Currently, most/all date fields defined on mongoose models are defined as:

```js
created: {
	type: Date,
	default: () => Date.now(),
	get: util.dateParse
}
```

The getter converts the date value to a numeric epoch value vs. an actual JS Date object.  When combined with use of the GetterPlugin, date values are also returned the same in JSON returned to client.  Without the getters, Dates in returned JSON would be converted to an ISO 8601 string (i.e. `2022-08-31T17:39:20.824Z`).

I did some quick testing on the User and Feedback models and removing the getter doesn't seem to cause any issues.  The UI still handles rendering the dates properly.

Another added benefit of removing the getters, is that it will allow us to use mongoose's  [timestamps feature](https://mongoosejs.com/docs/timestamps.html) option.  When using this, mongoose will automatically create the created and updated fields on schema as well as automatically handle setting values when appropriate.  This will be especially helpful for updates as currently we are manually setting a new updated date.

### PR DESCRIPTION
**Pull request by eacaps**
Closes #207

### CODE REVIEW

File: src/app/core/notifications/notification.model.js [hunk 1]
```diff
@@ -19,8 +24,6 @@ const NotificationSchema = new mongoose.Schema(
 		},
 		created: {
 			type: Date,
-			default: () => Date.now(),
-			get: util.dateParse,
 			expires: config.notificationExpires
```
Comment (eacaps) [inline src/app/core/notifications/notification.model.js:27]: While testing to make sure `expires` still works when using `timestamps` I came across a potential issue. If the expires value is changed in the config that will mess up the index the next time the server is run:
```
[2022-09-13T14:17:17.298Z]  INFO: application/48682 on ecapito: Mongoose: Setting debug to false
[2022-09-13T14:17:17.641Z]  INFO: application/48682 on ecapito: Starting initialization of Node.js server
[2022-09-13T14:17:17.703Z]  INFO: application/48682 on ecapito: Mongoose: Connected to "admin" default db
[2022-09-13T14:17:17.936Z] FATAL: application/48682 on ecapito: Mongoose: Could not connect to admin db
[2022-09-13T14:17:17.997Z] FATAL: application/48682 on ecapito: Startup initialization failed. (err.code=85)
    MongoServerError: Index with name: created_1 already exists with different options
        at Connection.onMessage (/Users/ecapito/workspace/node-rest-starter/node_modules/mongoose/node_modules/mongodb/src/cmap/connection.ts:438:20)
        at MessageStream.<anonymous> (/Users/ecapito/workspace/node-rest-starter/node_modules/mongoose/node_modules/mongodb/src/cmap/connection.ts:256:56)
        at MessageStream.emit (node:events:513:28)
        at processIncomingData (/Users/ecapito/workspace/node-rest-starter/node_modules/mongoose/node_modules/mongodb/src/cmap/message_stream.ts:193:14)
        at MessageStream._write (/Users/ecapito/workspace/node-rest-starter/node_modules/mongoose/node_modules/mongodb/src/cmap/message_stream.ts:70:5)
        at writeOrBuffer (node:internal/streams/writable:391:12)
        at _write (node:internal/streams/writable:332:10)
        at MessageStream.Writable.write (node:internal/streams/writable:336:10)
        at Socket.ondata (node:internal/streams/readable:754:22)
        at Socket.emit (node:events:513:28)
```

I changed the expires to `1m` to make sure things were still aging off. This broke and the server would not start until I ran `db.notifications.drop()`. I did verify `expires` still works with timestamps tho 🤷‍♂️

Comment (jrassa) [reply src/app/core/notifications/notification.model.js:27]: This is known behavior with mongoose.  You could also have fixed by just dropping the changed index using `db.collection.dropIndex(indexName)`.  Mongoose also has a [syncIndex](https://thecodebarbarian.com/whats-new-in-mongoose-5-2-syncindexes) function that could be used to resolve this.

**Review by jrassa — CHANGES_REQUESTED** [review]
Need to verify any internal uses of these date fields still function properly.  I did a quick check and found only one issue.

Two comparisons using `User.lastLogin` in [proxy-pki.js](https://github.com/bluehalo/node-rest-starter/blob/207-remove-getter-on-date-fields-in-mongoose-models/src/lib/strategies/proxy-pki.js#L89) likely need to be reworked. 

```js
if (
	!secondaryUser.lastLogin ||
	secondaryUser.lastLogin + config.auth.sessionCookie.maxAge < Date.now()
) {
}
```

**Review by jrassa — CHANGES_REQUESTED** [review]
With the switch to using the mongoose timestamps feature, all of the services can be updated to removing the explicit setting of `created` and `updated` values.  I don't think their presence is an issue as any values being set will be ignored, but we should clean it up to remove them.

For example: in `eua.service`:
```js
/**
 * @param obj
 * @returns {Promise<UserAgreementDocument>}
 */
create(obj) {
	const document = new this.model(obj);
	document.created = Date.now();
	document.updated = document.created;
	return document.save();
}
```
Can be replaced with:
```js
/**
 * @param obj
 * @returns {Promise<UserAgreementDocument>}
 */
create(obj) {
	const document = new this.model(obj);
	return document.save();
}
```

**Comment by jrassa** [comment]
Also need to verify any usages within ngx-starter.

This check in `authorization.service.ts` likely needs to be reworked.

```ts
public isEuaCurrent() {
	const euaPublished: number = this.session?.user?.eua?.published ?? 0;
	const euaAccepted: number = this.session?.user?.userModel?.acceptedEua ?? 0;

	return euaAccepted >= euaPublished;
}
```

File: src/spec/helpers.js [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+	return JSON.parse(JSON.stringify(json));
+};
+
+/**
+ * according to the internet this is a legit way to check if a string is
+ * an iso string
+ * @param {*} val
+ * @returns
+ */
+module.exports.isISOString = (val) => {
```
Comment (brianghig) [inline src/spec/helpers.js:20]: @eacaps - can we get some test coverage for this specific method? It may be covered by the model tests, but would be good to know how it can + should handle specific edge case inputs.

File: src/spec/helpers.js [hunk 3]
```diff
@@ -0,0 +1,21 @@
+/**
+ * this takes an object, converts stringifies it and then parses it back out.
+ * this is useful when fields are Date objects and need to be compared to a
+ * stringified version
+ * @param {*} json
+ * @returns
+ */
+module.exports.parsedJSON = (json) => {
+	return JSON.parse(JSON.stringify(json));
```
Comment (brianghig) [inline src/spec/helpers.js:11]: @eacaps - should we add any null-ish error handling here, so null doesn't get converted to "null" ? Might be overkill for the well-defined use case in the method docs, but it would be nice to get some test coverage for this.

File: src/spec/helpers.js (hunk 2 above)
Comment (eacaps) [reply src/spec/helpers.js:20]: The `isISOString` and `parsedJSON` methods are only supposed to be used within the context of specs. Ideally they could be injected into mocha some way in `src/test.js` rather than specifying them in a separate file but I couldn't find a clean way to do that. In Firehawk I also moved `clearDatabase` and `userSpec` to a [helper file](https://bitbucket.org/asymmetrik/fh-webapp/src/a7b67c560e753267e87c1a00d11ad37aa71ec7ad/src/server/app/util/spec/helpers.js#lines-118) so that code is not duplicated in multiple tests.

Comment (eacaps) [reply src/spec/helpers.js:20]: To add on here, I wouldn't be opposed to writing some tests to make sure these helper methods do what they say they do, but I just wanted to clarify their purpose.

File: src/app/core/audit/audit.model.js [hunk 4]
```diff
@@ -18,25 +16,28 @@ const mongoose = require('mongoose'),
...
-		userSpec: {
-			browser: { type: String },
-			os: { type: String }
-		},
-		masqueradingUser: { type: String }
+	{
+		timestamps: {
+			createdAt: 'created',
+			updatedAt: 'updated'
+		}
```
Comment (jrassa) [inline src/app/core/audit/audit.model.js]: Schema didn't have an `updated` field since they are never updated.  The updated timestamp can be disabled by settings `updatedAt` to false.

File: src/app/core/export/export-config.model.js [hunk 5]
```diff
@@ -2,31 +2,33 @@
...
-		type: Date,
-		expires: 300,
-		default: () => Date.now(),
-		get: utilService.dateParse,
-		required: [true, 'Created date is required']
+	{
+		timestamps: {
+			createdAt: 'created',
+			updatedAt: 'updated'
+		}
```
Comment (jrassa) [inline src/app/core/export/export-config.model.js]: Schema didn't have an `updated` field since they are never updated.  The updated timestamp can be disabled by settings `updatedAt` to false.

File: src/app/core/messages/dismissed-message.model.js [hunk 6]
```diff
@@ -15,21 +14,24 @@ const mongoose = require('mongoose'),
...
-	},
-	created: {
-		type: Date,
-		default: () => Date.now(),
-		get: util.dateParse
+	{
+		timestamps: {
+			createdAt: 'created',
+			updatedAt: 'updated'
+		}
```
Comment (jrassa) [inline src/app/core/messages/dismissed-message.model.js]: Schema didn't have an `updated` field since they are never updated.  The updated timestamp can be disabled by settings `updatedAt` to false.

File: src/app/core/notifications/notification.model.js [hunk 7]
```diff
@@ -2,13 +2,18 @@
...
 /**
  * Notification Schema
  */
-module.exports.notificationOptions = { discriminatorKey: 'notificationType' };
+module.exports.notificationOptions = {
+	discriminatorKey: 'notificationType',
+	timestamps: {
+		createdAt: 'created',
+		updatedAt: 'updated'
+	}
```
Comment (jrassa) [inline src/app/core/notifications/notification.model.js]: Schema didn't have an `updated` field since they are never updated.  The updated timestamp can be disabled by settings `updatedAt` to false.

**Review by jrassa — CHANGES_REQUESTED** [review]

File: src/spec/helpers.js (hunk 2 above)
Comment (eacaps) [reply src/spec/helpers.js:20]: I went ahead and added some small specs for the helpers

**Comment by eacaps** [comment]
I created a related PR for ngx-starter: https://github.com/bluehalo/ngx-starter/pull/283

**Review by jrassa — APPROVED** [review]
